Write release notes for occlusion culling (#2072)
bevyengine/bevy-website: release-content/0.16/release-notes/occlusion_culling.md

> When it comes to pushing hardware to support larger levels and more detailed meshes, "simply draw less stuff" is a tried-and-true strategy.

Reviewer: Swap levels for scenes, given that’s what we call them?

> These techniques are collectively referred to as "culling": we can save work whenever we can determine "this doesn't need to be drawn" for cheaper than it would cost to draw the thing despite its pixels later being overwritten completely.

Reviewer 2: ```suggestion
These techniques are collectively referred to as "culling": we can save work whenever we can determine "this doesn't need to be drawn" for cheaper than it would cost to actually perform the draw.
```

**Occlusion culling** is the idea that we don't need to draw something that's completely blocked by other opaque objects,
from the perspective of the camera.

> That makes sense! We don't need to draw a person hidden behind a wall, even if they're within the range used for frustrum culling.

Author: ```suggestion
That makes sense! We don't need to draw a person hidden behind a wall, even if they're within the range used for frustum culling.
```

> That makes sense! We don't need to draw a car parked on the other side of a building, even if it's within the range used for fustrum culling.

Reviewer 2: ```suggestion
That makes sense! We don't need to draw a person hidden standing a wall, even if they're within the range used for frustum culling.
```

Reviewer 2: person hidden behind a wall*

Author: ```suggestion
That makes sense! We don't need to draw a person hidden behind a wall, even if it's within the range used for frustrum culling.
```

> That makes sense! We don't need to draw a car parked on the other side of a building (even if it's within the range used for fustrum culling), and we *never* need to draw all of the teeth inside of a character's closed mouth.

Reviewer 3: I don't know if the teeth mentioned here is the best example. Most of the time, they will just be part of the character mesh and unless we are talking about meshlets, it's generally the entire character mesh that is completely visible or not.

Maybe something about vegetation on the other side of a mountain? Or trees occluding other trees in a dense forest?

Before you worry too much: we're already checking for this category of wasted work in some form, via the use of a **depth prepass**.

> A depth prepass renders a simplified version of the scene, recording only the depth (distance from the camera) of each object into the 2-dimensional depth buffer.

Reviewer 3: Saying "a simplified version of the scene" kinda implies to me that it uses simplified meshes, but that's not what's happening. I think it would be more accurate and less misleading to only mention the depth only part without saying it's a simplified scene. Maybe saying it's a simplified shader?

Author: ```suggestion
A depth prepass renders the scene with a simplified shader, recording only the depth (distance from the camera) of each object into the 2-dimensional depth buffer.
```

> We can check the resulting depth buffer, count how many pixels are associated with each object, and then discard objects whose pixels were completely overwritten.

Reviewer 3: That's not really what the depth prepass is doing. We don't do any kind of pixel counting (unless I'm misunderstanding what you mean by that). What it does is record the top most fragment for each fragment on the screen, then during the main pass we can use that information and avoid any work that would be behind that top level fragment.

Reviewer: I don’t think of it as the ’topmost’ but rather the closest opaque fragment to the camera along the ray through the centre of that fragment.

Because of this approach, we can avoid most fragment shading costs (dominated by textures and lighting) for occluded objects, but the vertex shading overhead is still present, in addition to the inherent cost of this fragment testing process.

We can do better than that.

> As [PR #17413] lays out, we're adopting the modern two-phase occlusion culling (in contrast to a traditional potentially visible sets design) already used by our virtual geometry rendering, because it works well with the GPU-driven rendering architecture that we've established during this cycle!

Reviewer 2: ```suggestion
As [PR #17413] lays out, we're adopting the modern two-phase occlusion culling (in contrast to a traditional potentially visible sets design) already used by our virtual geometry rendering system, because it works well with the GPU-driven rendering architecture that we've established during this cycle!
```

> As [PR #17413] lays out, we're adopting the modern two-phase occlusion culling (in contrast to a traditional potentially visible sets design) already used by our virtual geometry rendering, because it works well with the GPU-driven rendering architecture (cold specialization, retained bins) that we've established during this cycle!

Author: I've cited 17413 here because the PR description does a great job explaining the gory technical details for curious rendering engineers.

> As [PR #17413] lays out, we're adopting the modern two-phase occlusion culling (in contrast to a traditional potentially visible sets design) already used by our virtual geometry rendering, because it works well with the GPU-driven rendering architecture (cold specialization, retained bins) that we've established during this cycle!

Reviewer 3: Cold specialization isn't related to GPU driven. It's an important part optimization for the renderer but it's relevant to both the cpu driven and gpu driven path. Same for retained bin actually. I would simply remove that parenthesis to be honest.

Reviewer 3: ```suggestion
As [PR #17413] lays out, we're adopting the modern two-phase occlusion culling (in contrast to a traditional potentially visible sets design) already used by our virtual geometry rendering, because it works well with the GPU-driven rendering architecture that we've established during this cycle!
```

> For now, this feature is marked as experimental, due to known precision issues that can mark meshes as occluded even when they're not.

Reviewer 3: Maybe mention that this could result in pop in. It might be easier to understand the tradeoff for a lot of people.

Reviewer 2: It's not really pop in, it'll be meshes not rendering when they're supposed to.

Reviewer 3: Sure, but that will just last a few frames and it will eventually get rendered again which will cause pop in right?

Reviewer 2: Not entirely sure, but don't think so. This is about over-agressive culling. If you shift the camera it might render again, though.

In practice, we're not convinced that this is a serious concern, so please let us know how it goes!
To try out the new mesh occlusion culling, add the [`DepthPrepass`] and [`OcclusionCulling`] components to your camera.

As we said above, the check has to be faster than simply drawing the thing: occlusion culling won't be faster on all scenes.
Small scenes, or those using simpler non-PBR rendering are particularly likely to be slower with occlusion culling turned on.
Like always: you need to measure your performance to improve it.

> If you're a rendering engineer who'd like to help us resolve these precision issues and stabilize this feature, we're looking to learn from Hans-Kristian Arntzen design in [Granite].

Reviewer 2: ```suggestion
If you're a rendering engineer who'd like to help us resolve these precision issues and stabilize this feature, we're looking to borrow from Hans-Kristian Arntzen's design in [Granite].
```
